### PR TITLE
Refactor collections services to resource services with single access

### PR DESCRIPTION
**Pull request by oliverguenther**
Refactors collection services into resource services, providing access to both collections and entities through their href.

By that, fixes https://community.openproject.org/work_packages/47518/activity by providing a way to require entities without loading them twice.

### CODE REVIEW

File: frontend/src/app/core/state/resource-store.ts [hunk 1]
```diff
@@ -36,58 +36,48 @@ export function mapHALCollectionToIDCollection<T extends CollectionItem>(collect
...
- * @param collectionUrl The key to insert the collection at
- * @param loading The loading state
+ * @param string The resource path to mark as loading
  */
-export function setCollectionLoading<T extends { id:ID }>(
-  store:EntityStore<CollectionState<T>>,
-  collectionUrl:string,
+export function setResourceLoading<T extends { id:ID }>(
+  store:EntityStore<ResourceState<T>>,
+  string:string,
```
Comment (b12f) [inline frontend/src/app/core/state/resource-store.ts]: Should probably be called `url` or similar

File: frontend/src/app/core/state/resource-store.service.ts [hunk 2]
```diff
@@ -87,42 +92,61 @@ export abstract class ResourceCollectionService<T extends { id:ID }> {
...
       );
   }
 
+  /**
+   * Require a single entity to be loaded.
+   * Returnes the cached entity if it was loaded already
+   *
+   * @param href {string}
+   */
+  public requireEntity(href:string):Observable<T> {
```
Comment (Kharonus) [inline frontend/src/app/core/state/resource-store.service.ts:118]: I just merged that branch into my rework of the storage component. and there I have noticed an effect of `requireEntity`. It must be the same as with `requireCollection`, but apparently we do not have a case yet.

So, in `ngOnInit()` I want to initialize my storage data:

```typescript
this.storage = this.storagesResourceService.requireEntity(this.projectStorage._links.storage.href);
```

the, this observable is used in a vast variety of places. In the template, in the component code fetching data before displaying texts, etc.

BUT: The observable behind it is the `fetchEntity()`, not the `lookup()`. Hence, every subscription to the storage observable executes a new HTTP request.

to circumvent it, I would have to write something like this:

```typescript
    this.storagesResourceService.requireEntity(this.projectStorage._links.storage.href).subscribe();
    this.storage = this.storagesResourceService.lookup(idFromLink(this.projectStorage._links.storage.href));
```
But from my POV this is something, that should be here in the resource store service. What are your opinions on that?

**Review by Kharonus — CHANGES_REQUESTED** [review]

File: frontend/src/app/core/state/resource-store.service.ts (hunk 2 above)
Comment (oliverguenther) [reply frontend/src/app/core/state/resource-store.service.ts:118]: Yes I agree, the outer interface should allow multiple subscribes transparently. In fact I even think an internal double subscription to the same fetch should be shared. Added that to both entity and collection here: https://github.com/opf/openproject/commit/2d09306a22

Comment (Kharonus) [reply frontend/src/app/core/state/resource-store.service.ts:118]: Perfect!

**Review by Kharonus — APPROVED** [review]
LGTM

File: frontend/src/app/core/state/resource-store.ts (hunk 1 above)
Comment (oliverguenther) [reply frontend/src/app/core/state/resource-store.ts]: Good catch, this wasn't really meant to be called string. Fixed that + the occurence below in removeResourceLoading

**Pull request merged**